File: tree/ordered_traversal/114_flatten_binary_tree_to_linked_list.cc
#include <stack>
#include <vector>

#include "tree_node.h"

// 2024/01/09: 看答案写出了前序遍历, 注意left和right都要重新赋值.
// TODO: 学习一下前序遍历和展开同时进行.
void Flatten(TreeNode* root) {
  std::vector<TreeNode*> preordered_traversal;
  PreorderedTraversal(preordered_traversal, root);
  for (int i = 1; i < preordered_traversal.size(); ++i) {
    TreeNode* prev = preordered_traversal[i - 1];
    TreeNode* curr = preordered_traversal[i];
    prev->left = nullptr;
    prev->right = curr;
  }
}

void PreorderedTraversal(std::vector<TreeNode*>& preordered_traversal,
                         TreeNode* root) {
  std::stack<TreeNode*> s;
  while (!s.empty() || root) {
    while (root) {
      preordered_traversal.push_back(root);
      s.push(root);
      root = root->left;
    }
    root = s.top();
    s.pop();
    root = root->right;
  }
}

File: tree/ordered_traversal/114_flatten_binary_tree_to_linked_list.py
from typing import Optional
from tree.tree_node import TreeNode, create_tree_node, print_tree_node


# 2026/05/19: 自己写出来, 前序遍历的同时展开.
def flatten(root: Optional[TreeNode]) -> None:
    """
    Do not return anything, modify root in-place instead.
    """
    if root is None:
        return

    stack = []
    prev = TreeNode(0)
    stack.append(root)
    while len(stack) > 0:
        curr = stack.pop()
        prev.right = curr
        prev.left = None
        prev = curr
        if curr.right is not None:
            stack.append(curr.right)
        if curr.left is not None:
            stack.append(curr.left)


root = [1, 2, 5, 3, 4, None, 6]
root = create_tree_node(root)
print_tree_node(root)
flatten(root)
print_tree_node(root)
